src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,113 @@
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        CancellationToken cancellationToken)
+    => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => TryResolveInsertionAsync(
+                context,
+                linePosition,
+                character,
+                autoCloseTags,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> TryResolveInsertionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        CancellationToken cancellationToken)
+    {
+        var sourceText = await remoteDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var index))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var codeDocument = await remoteDocumentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        var languageKind = _documentMappingService.GetLanguageKind(codeDocument, index, rightAssociative: true);
+        if (languageKind is RazorLanguageKind.Html)
+        {
+            return Response.CallHtml;
+        }
+        else if (languageKind is RazorLanguageKind.Razor)
+        {
+            linePosition.Deconstruct(out var line, out var lineCharacter);
+            var insertTextEdit = await _autoInsertService.TryResolveInsertionAsync(
+                remoteDocumentContext.Snapshot,
+                new Position(line, lineCharacter),
+                character,
+                autoCloseTags,
+                cancellationToken);
+            return insertTextEdit.HasValue
+                ? Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(insertTextEdit.Value))
+                : Response.NoFurtherHandling;
+        }
+
+        // C# case
+
+        var csharpDocument = codeDocument.GetCSharpDocument();
+        if (_documentMappingService.TryMapToGeneratedDocumentPosition(csharpDocument, index, out var mappedPosition, out _))
+        {
+            var generatedDocument = await remoteDocumentContext.GetGeneratedDocumentAsync(_filePathService, cancellationToken).ConfigureAwait(false);
+            // TODO: use correct options rather than default
+            var formattingOptions = new RoslynFormattingOptions();
+            var autoInsertResponseItem = await OnAutoInsert.GetOnAutoInsertResponseAsync(
+                generatedDocument,
+                mappedPosition,
+                character,
+                formattingOptions,
+                cancellationToken
+            );
+            return autoInsertResponseItem is not null
+                ? Response.Results(RemoteInsertTextEdit.FromRoslynAutoInsertResponse(autoInsertResponseItem))
```

Comment:
Yes unfortunately. It's done in the original endpoint by the RazorFormattingService. I didn't realize that besides doing code formatting it also remaps document positions from C# to Razor (during CSharpOnTypeFormattingPass from what I can tell). 

Does it mean that we need to move RazorFormattingService to the common layer after all?

Reply:
Ahh yeah, I forgot that on type formatting (which this would use internally) works on being given a set of C# edits. Yes, this will need the formatting service to move down.